File: simulator/src/message/TestMessage.cpp

#include "Message.h"
#include "MessageType.h"
#include "TestMessage.h"
#include "TestMessageId.h"
#include "TestResponseId.h"

TestMessage::TestMessage() : Message()
{

}

TestMessage::TestMessage( string message )
           : Message( MessageHeader::getHeaderSize() + message.length() + MessageTrailer::getTrailerSize(),
                      TEST_MESSAGE, TEST_MESSAGE_STRING, TEST_MESSAGE_VERSION )
{
    this->message = message;
}

TestMessage::~TestMessage()
{
}

string 
TestMessage::getMessage()
{
    return message;
}

TestMessageResponse *
TestMessage::processMessage()
{
    return new TestMessageResponse( TEST_RESPONSE_SUCCESS,
                                    "Received message: " + message );
}

void
TestMessage::pack( RawBuffer *buffer )
{
    buffer->put( (char *)message.c_str(), message.length() );
}

TestMessage *
TestMessage::unpack( RawBuffer *buffer )
{
    TestMessage *tm = new TestMessage();
    tm->header  = MessageHeader::unpack( buffer );
    tm->trailer = MessageTrailer::unpack( buffer );

    int dataLength = tm->header->getLength() - MessageHeader::getHeaderSize()
                   - MessageTrailer::getTrailerSize();
    char *data = new char[dataLength+1];
    buffer->setPosition( MessageHeader::getHeaderSize() );
    buffer->get( data, 0, dataLength );
    data[dataLength] = '\0';

    tm->message.append( data );

    delete data;

    return tm;
}

File: interface/src/com/christopher/robot/message/TestMessage.java
package com.christopher.robot.message;

import java.nio.ByteBuffer;
import java.nio.charset.Charset;

public class TestMessage extends Message
{
    private static int TEST_MESSAGE_VERSION = 0;
    private String message;

    private TestMessage()
    {
        super();
    }

    public TestMessage( TestMessageId id, String message )
    {
        super( message.length(), MessageType.TEST_MESSAGE.getValue(),
               id.getId(), TEST_MESSAGE_VERSION );
        this.message = message;
    }

    public String getMessage()
    {
        return message;
    }

    private void setMessage( String message )
    {
        this.message = message;
    }

    public Response processMessage()
    {
        System.out.println( "Received: " + message );
        return new TestMessageResponse( TestResponseId.TEST_RESPONSE_SUCCESS,
                                        "Received message: " + message );
    }

    public void pack( ByteBuffer buffer )
    {
        header.pack( buffer );
        buffer.put( message.getBytes( Charset.forName( "US-ASCII") ) );
        trailer.pack( buffer );
    }

    public static TestMessage unpack( ByteBuffer buffer ) throws MessageException
    {
        TestMessage tm = new TestMessage();

        tm.header = MessageHeader.unpack( buffer );
        tm.trailer = MessageTrailer.unpack( buffer );

        int dataLength = tm.header.getLength() - MessageHeader.getSize()
                       - MessageTrailer.getSize();
        byte[] data = new byte[dataLength];
        buffer.get( data, MessageHeader.getSize(), dataLength );

        tm.setMessage( new String( data, Charset.forName( "US-ASCII" ) ) );

        return tm;
    }
}
